### PR TITLE
changes for inference methods

### PR DESCRIPTION
**Pull request by palashahuja**

### CODE REVIEW

**Comment by ankurankan** [comment]
@palashahuja Tests are failing.. Can you please check?

**Comment by palashahuja** [comment]
@ankurankan The test cases are passing now. Should I try merging the methods `get_cpds` and `get_cpds_by_slice`?

**Comment by ankurankan** [comment]
@palashahuja Yes you should merge the two with a default value for the time slice parameter.

**Comment by ankurankan** [comment]
@palashahuja I don't want the methods to throw error when the time slice is greater than 1. Simply construct the node, edge or CPD according to that slice and return it.

File: pgmpy/models/DynamicBayesianNetwork.py [hunk 1]
```diff
@@ -183,9 +183,13 @@ def add_edges_from(self, ebunch, **kwargs):
         for edge in ebunch:
             self.add_edge(edge[0], edge[1])
 
-    def get_intra_edges(self):
+    def get_intra_edges(self, time_slice=0):
         """
         returns the intra slice edges present in the 2-TBN.
+        Parameter
+        ---------
+        time_slice: The timeslice should be a positive value greater than or equal to zero
```
Comment (ankurankan) [inline pgmpy/models/DynamicBayesianNetwork.py]: @palashahuja Please check the way docstrings have been written. First line should be the data type and second line should be the description of the argument.

File: pgmpy/models/DynamicBayesianNetwork.py [hunk 2]
```diff
@@ -211,6 +218,47 @@ def get_inter_edges(self):
         """
         return [edge for edge in self.edges() if edge[0][1] != edge[1][1]]
 
+    def get_interface_nodes(self, time_slice=0):
+    	"""
+    	returns the nodes in the first timeslice whose children are present in the first timeslice.
+        Parameter
+        ---------
+        time_slice: The timeslice should be a positive value greater than or equal to zero
```
Comment (ankurankan) [inline pgmpy/models/DynamicBayesianNetwork.py]: Same here

File: pgmpy/models/DynamicBayesianNetwork.py [hunk 3]
```diff
@@ -195,7 +199,10 @@ def get_intra_edges(self):
         >>> dbn.get_intra_edges()
         [(('D', 0), ('G', 0)), (('G', 0), ('L', 0)), (('I', 0), ('G', 0))
         """
-        return [edge for edge in self.edges() if edge[0][1] == edge[1][1]]
+        if not str(time_slice).isdigit():
```
Comment (ankurankan) [inline pgmpy/models/DynamicBayesianNetwork.py]: @palashahuja What is this? A better way to do this check would be `if not isinstance(time_slice, int) or time_slice < 0`

File: pgmpy/models/DynamicBayesianNetwork.py [hunk 4]
```diff
@@ -211,6 +218,47 @@ def get_inter_edges(self):
...
+    	Examples:
+    	-------
+    	>>> from pgmpy.models import DynamicBayesianNetwork as DBN
+    	>>> dbn = DBN()
+    	>>> dbn.add_nodes_from(['D', 'G', 'I', 'S', 'L'])
+    	>>> dbn.add_edges_from([(('D',0),('G',0)),(('I',0),('G',0)),(('G',0),('L',0)),(('D',0),('D',1))])
+    	>>> dbn.get_interface_nodes()
+    	[('D', 0)]
+    	"""
+    	if not str(time_slice).isdigit():
```
Comment (ankurankan) [inline pgmpy/models/DynamicBayesianNetwork.py]: Same here

File: pgmpy/models/DynamicBayesianNetwork.py [hunk 5]
```diff
@@ -211,6 +218,47 @@ def get_inter_edges(self):
...
+
+    	Examples:
+    	-------
+    	>>> from pgmpy.models import DynamicBayesianNetwork as DBN
+    	>>> dbn = DBN()
+    	>>> dbn.add_nodes_from(['D', 'G', 'I', 'S', 'L'])
+    	>>> dbn.add_edges_from([(('D',0),('G',0)),(('I',0),('G',0)),(('G',0),('L',0)),(('D',0),('D',1))])
+    	>>> dbn.get_slice_nodes()
+    	"""
+    	if not str(time_slice).isdigit():
```
Comment (ankurankan) [inline pgmpy/models/DynamicBayesianNetwork.py]: Same here

File: pgmpy/tests/test_models/test_DynamicBayesianNetwork.py [hunk 6]
```diff
@@ -69,17 +69,40 @@ def setUp(self):
...
         self.network.add_edges_from([(('a', 0), ('b', 0)), (('a', 0), ('a', 1)), (('b', 0), ('b', 1))])
-        self.assertListEqual(sorted(self.network.get_intra_edges()), [(('a', 0), ('b', 0)), (('a', 1), ('b', 1))])
+        self.assertListEqual(sorted(self.network.get_intra_edges()), [(('a', 0), ('b', 0))])
+        self.assertListEqual(sorted(self.network.get_intra_edges(1)), [(('a', 1), ('b', 1))])
         self.assertListEqual(sorted(self.network.get_inter_edges()), [(('a', 0), ('a', 1)), (('b', 0), ('b', 1))])
 
+    def test_get_interface_nodes(self):
+    	self.network.add_edges_from(
+    		[(('D', 0), ('G', 0)), (('I', 0), ('G', 0)), (('D', 0), ('D', 1)), (('I', 0), ('I', 1))])
+    	self.assertListEqual(sorted(self.network.get_interface_nodes()), [('D', 0), ('I',0)])
```
Comment (ankurankan) [inline pgmpy/tests/test_models/test_DynamicBayesianNetwork.py]: @palashahuja Add tests for other time slices and errors.

File: pgmpy/tests/test_models/test_DynamicBayesianNetwork.py [hunk 7]
```diff
@@ -69,17 +69,40 @@ def setUp(self):
...
+
+    def test_get_slice_nodes(self):
+    	self.network.add_edges_from(
+    		[(('D', 0), ('G', 0)), (('I', 0), ('G', 0)), (('D', 0), ('D', 1)), (('I', 0), ('I', 1))])
+    	self.assertListEqual(sorted(self.network.get_slice_nodes()), [('D', 0), ('G', 0), ('I', 0)])
+    	self.assertListEqual(sorted(self.network.get_slice_nodes(1)), [('D', 1), ('G', 1), ('I', 1)])
```
Comment (ankurankan) [inline pgmpy/tests/test_models/test_DynamicBayesianNetwork.py]: Add tests for errors.

**Comment by ankurankan** [comment]
@palashahuja Please check the comments. Everything else looks fine.

**Comment by palashahuja** [comment]
@ankurankan `str(x).isdigit()` will always check for positive numbers greater than or equal to 0. Should I change it ?

**Comment by ankurankan** [comment]
Doesn't seem a very good way to check to me as its not very readable.

**Pull request merged**